drivers/resolver/ad2s1210/ad2s1210.c

```diff
@@ -0,0 +1,408 @@
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <stdlib.h>
+#include "ad2s1210.h"
+#include "no_os_util.h"
+#include "no_os_error.h"
+#include "no_os_print_log.h"
+#include "no_os_delay.h"
+#include "no_os_alloc.h"
+
+static int32_t ad2s1210_set_mode_pins(struct ad2s1210_dev *dev,
+				      enum ad2s1210_mode mode);
+
+/*******************************************************************************
+* @brief Write to device register
+*
+* @param dev - Device descriptor.
+* @param addr - Address of the register to write
+* @param val - value to write
+*
+* @return Returns negative error code or 0 in case of success.
+*******************************************************************************/
+int32_t ad2s1210_reg_write(struct ad2s1210_dev *dev, uint8_t addr, uint8_t val)
+{
+	uint8_t buf;
+	int ret;
+
+	if (addr < AD2S1210_REG_MIN)
+		return -EINVAL;
+
+	ret = ad2s1210_set_mode_pins(dev, MODE_CONFIG);
+	if (ret)
+		return ret;
+
+	buf = addr;
+	ret = no_os_spi_write_and_read(dev->spi_desc, &buf, 1);
+	if (ret)
+		return ret;
+
+	buf = val;
+	return no_os_spi_write_and_read(dev->spi_desc, &buf, 1);
+}
+
+/*******************************************************************************
+* @brief Read a device register
+*
+* @param dev - Device descriptor.
+* @param addr - Address of the register to read
+* @param val - pointer to store register value
+*
+* @return Returns negative error code or 0 in case of success.
+*******************************************************************************/
+int32_t  ad2s1210_reg_read(struct ad2s1210_dev *dev, uint8_t addr, uint8_t *val)
+{
+	uint8_t buf;
+	int32_t ret;
+
+	if (addr < AD2S1210_REG_MIN)
+		return -EINVAL;
+
+	ret = ad2s1210_set_mode_pins(dev, MODE_CONFIG);
+	if (ret)
+		return ret;
+
+	buf = addr;
+	ret = no_os_spi_write_and_read(dev->spi_desc, &buf, 1);
+	if (ret)
+		return ret;
+
+	buf = addr;
+	ret = no_os_spi_write_and_read(dev->spi_desc, &buf, 1);
```

Comment:
this operation seems duplicated

Reply:
the first operation will write the address.. the second operation is to read the data. we need to toggle CS between the two so that is why there are 2 separate operations.

Reply:
got it. Then is the second `buff = addr` assignment required? (at line 107)

Reply:
from the data sheet (p29) it seems that some valid address needs to be on SDI  pins. I saw the chip stop responding if we try to read some invalid address like 0x00. Actually while we read data on the SDO pin the chip prepares the data from the address that it sees in the SDI pins. this is for multiple byte reads. So better to make sure buff a valid address and not data from the prev read.

Reply:
got it, thanks! can you please add a suggestive comment in the code regarding this a bit unusual handling of the spi transfers?

Reply:
yes ok.